diff --git a/cc/configure_tensorflow.sh b/cc/configure_tensorflow.sh
index 3bc9cea..2a5db5c 100755
--- a/cc/configure_tensorflow.sh
+++ b/cc/configure_tensorflow.sh
@@ -20,25 +20,27 @@ fi
 
 # TODO(tommadams): we should probably switch to Clang at some point.
 
-if [ -d "${tmp_dir}" ]; then
-  pushd "${tmp_dir}"
-else
-  echo "Cloning tensorflow to ${tmp_dir}"
+if [ ! -d "${tmp_dir}" ]; then
   git clone https://github.com/tensorflow/tensorflow "${tmp_dir}"
+  echo "Cloning tensorflow to ${tmp_dir}"
+fi
 
-  pushd "${tmp_dir}"
+pushd "${tmp_dir}"
 
-  cherry_pick_tag="02c111ab4269ab73a506164e4b54ba996d28a8cf"
-  prev_tag="8be9158c7a701d933bbe532f5d54df17f47a4284"
+git reset --hard
+git clean -fd
 
-  git diff "${prev_tag}" "${cherry_pick_tag}" > sample.patch
+cherry_pick_tag="02c111ab4269ab73a506164e4b54ba996d28a8cf"
+prev_tag="8be9158c7a701d933bbe532f5d54df17f47a4284"
+
+git diff "${prev_tag}" "${cherry_pick_tag}" > sample.patch
+
+commit_tag="961bb02b882a8bb921e5be1c09c34b51fffd25dc"
+echo "Checking out ${commit_tag}"
+git checkout "${commit_tag}"
+git apply sample.patch
+cp ${script_dir}/../ml_perf/tools/tensorflow_quantization/graph_transforms/fuse_quantized_convolution.cc ${tmp_dir}/tensorflow/tools/graph_transforms/
 
-  commit_tag="961bb02b882a8bb921e5be1c09c34b51fffd25dc"
-  echo "Checking out ${commit_tag}"
-  git checkout "${commit_tag}"
-  git apply sample.patch
-  cp ${script_dir}/../ml_perf/tools/tensorflow_quantization/graph_transforms/fuse_quantized_convolution.cc ${tmp_dir}/tensorflow/tools/graph_transforms/
-fi
 
 # Run the TensorFlow configuration script, setting reasonable values for most
 # of the options.
-- 
1.8.3.1

