diff --git a/cc/configure_tensorflow.sh b/cc/configure_tensorflow.sh
index 2a5db5c..854218f 100755
--- a/cc/configure_tensorflow.sh
+++ b/cc/configure_tensorflow.sh
@@ -49,6 +49,8 @@ cc_opt_flags="${CC_OPT_FLAGS:--march=native}"
 
 PYTHON_BIN_PATH=`which python`
 
+bazel clean
+
 CC_OPT_FLAGS="${cc_opt_flags}" \
 PYTHON_BIN_PATH=${PYTHON_BIN_PATH} \
 USE_DEFAULT_PYTHON_LIB_PATH="${USE_DEFAULT_PYTHON_LIB_PATH:-1}" \
-- 
1.8.3.1

