diff --git a/ml_perf/get_data.py b/ml_perf/get_data.py
index 680209f..6dd745b 100644
--- a/ml_perf/get_data.py
+++ b/ml_perf/get_data.py
@@ -49,24 +49,14 @@ def freeze_graph(path):
 
 
 def main(unused_argv):
-  try:
-    for d in ['checkpoint', 'target']:
-      # Pull the required training checkpoints and models from GCS.
-      src = os.path.join(FLAGS.src_dir, d, N)
-      dst = os.path.join(FLAGS.dst_dir, d)
-      utils.ensure_dir_exists(dst)
-      utils.wait(utils.checked_run('gsutil', '-m', 'cp', '-r', src, dst))
-
-    # Freeze the target model.
-    freeze_graph(os.path.join(FLAGS.dst_dir, 'target', N, 'target'))
-
-    # Freeze the training checkpoint models.
-    pattern = os.path.join(FLAGS.dst_dir, 'checkpoint', N, 'work_dir', '*.index')
-    for path in glob.glob(pattern):
-      freeze_graph(os.path.splitext(path)[0])
-
-  finally:
-    asyncio.get_event_loop().close()
+  # Freeze the target model.
+  freeze_graph(os.path.join(FLAGS.dst_dir, 'target', N, 'target'))
+
+  # Freeze the training checkpoint models.
+  pattern = os.path.join(FLAGS.dst_dir, 'checkpoint', N, 'work_dir', '*.index')
+  for path in glob.glob(pattern):
+    freeze_graph(os.path.splitext(path)[0])
+
 
 
 if __name__ == '__main__':
-- 
2.25.0

