diff --git a/run_minigo_mn.sh b/run_minigo_mn.sh
index 06e0633..63c713f 100755
--- a/run_minigo_mn.sh
+++ b/run_minigo_mn.sh
@@ -29,7 +29,7 @@ BOARD_SIZE=9  python3  ml_perf/reference_implementation.py \
   `ml_perf/hostlist.sh |head -n $PLAY_NODES_PLUS_ONE|tail -n $PLAY_NODES|awk '/./{print "--selfplay_node="$0}'` \
   `ml_perf/hostlist.sh |tail -n $TRAIN_NODES|awk '/./{print "--train_node="$0}'` \
   `ml_perf/hostlist.sh |tail -n $TRAIN_PLUS_EVAL_NODES|head -n $EVAL_NODES |awk '/./{print "--eval_node="$0}'` \
-  --setup_train_workers=True &> train_workers.log &
+  --setup_train_workers=True > train_workers.log
 
 echo "BOARD_SIZE=9  python3  ml_perf/reference_implementation.py"
 # Run training loop
-- 
1.8.3.1

